pyogrio/_io.pyx

```diff
@@ -2523,6 +2673,29 @@ def ogr_write_arrow(
             if array.release != NULL:
                 array.release(&array)
 
+        # close dataset to force driver to flush data
+        exc = ogr_close(ogr_dataset)
+        ogr_dataset = NULL
+        if exc:
+            raise DataSourceError(f"Failed to write features to dataset {path}; {exc}")
+
+        ### Copy in-memory file back to path_or_fp object
+        if vsi_handle != NULL:
+            # rewind to beginning of file and read bytes back from memory
+            VSIFSeekL(vsi_handle, 0, 0)
+
+            # Take ownership of the buffer to avoid a copy; GDAL will automatically
+            # unlink the memory file
+            vsi_buffer = VSIGetMemFileBuffer(path.encode("UTF-8"), &vsi_buffer_size, 1)
+            if vsi_buffer == NULL:
+                raise RuntimeError("could not read bytes from in-memory file")
+
+            # reset path_or_fp and write bytes to it
+            path_or_fp.seek(0)
```

Comment:
This is the user-provided BytesIO object at this point, right? 
In that case, should we seek into it? I would expect we just write to it as the user provided it (they might want to write to a specific place)

Reply:
Yes, it is the user-provided BytesIO object.  However, we don't allow it to have existing bytes, so this first seek is unnecessary.  I don't think we want to support allowing writes to arbitrary locations within a passed-in BytesIO object; if we ever allowed it to be non-empty (e.g., if we could get append working) then we'd want to completely overwrite it.